Userland/Libraries/LibMarkdown/List.cpp

```diff
@@ -38,96 +47,95 @@ String List::render_for_terminal(size_t) const
             builder.appendff("{}. ", ++i);
         else
             builder.append("* ");
-        builder.append(item.render_for_terminal());
+        builder.append(item->render_for_terminal());
         builder.append("\n");
     }
     builder.append("\n");
 
     return builder.build();
 }
 
-OwnPtr<List> List::parse(Vector<StringView>::ConstIterator& lines)
+OwnPtr<List> List::parse(LineIterator& lines)
 {
-    Vector<Text> items;
-    bool is_ordered = false;
+    Vector<OwnPtr<ContainerBlock>> items;
 
     bool first = true;
-    size_t offset = 0;
-    StringBuilder item_builder;
-    auto flush_item_if_needed = [&] {
-        if (first)
-            return true;
+    bool is_ordered = false;
 
-        auto text = Text::parse(item_builder.string_view());
-        items.append(move(text));
+    bool is_tight = true;
+    bool has_trailing_blank_lines = false;
+    size_t start_number = 1;
 
-        item_builder.clear();
-        return true;
-    };
+    while (!lines.is_end()) {
+
+        size_t offset = 0;
 
-    while (true) {
-        if (lines.is_end())
-            break;
         const StringView& line = *lines;
-        if (line.is_empty())
-            break;
 
         bool appears_unordered = false;
-        if (line.length() > 2) {
-            if (line[1] == ' ' && (line[0] == '*' || line[0] == '-')) {
+
+        while (offset < line.length() && line[offset] == ' ')
+            ++offset;
+
+        if (offset + 2 <= line.length()) {
+            if (line[offset + 1] == ' ' && (line[offset] == '*' || line[offset] == '-' || line[offset] == '+')) {
                 appears_unordered = true;
-                offset = 2;
+                offset++;
             }
         }
 
         bool appears_ordered = false;
-        for (size_t i = 0; i < 10 && i < line.length(); i++) {
+        for (size_t i = offset; i < 10 && i < line.length(); i++) {
             char ch = line[i];
             if ('0' <= ch && ch <= '9')
                 continue;
             if (ch == '.' || ch == ')')
                 if (i + 1 < line.length() && line[i + 1] == ' ') {
+                    auto maybe_start_number = line.substring_view(offset, i - offset).to_uint<size_t>();
```

Comment:
Also, looks like this specialisation of `to_uint` doesn't exist.
These two silly declarations were probably meant to be marked `unsigned` (in StringView.cpp:230):

```c++
template Optional<long> StringView::to_uint() const;
template Optional<long long> StringView::to_uint() const;
```

Reply:
This is very weird, see my commit message for the last commit.

Reply:
also the integer instantiations are actually used in libline, so i'm leaving them.